dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/programrule/ProgramRuleTest.java

```diff
@@ -265,129 +269,265 @@ void shouldNotImportWhenErrorIsTriggeredOnAllEntities()
+
+        ImportReport importReport = trackerImportService.importTracker( params );
+
         assertNoErrors( importReport );
-        warningReports = importReport.getValidationReport().getWarnings();
-        assertEquals( 1, warningReports.size() );
-        assertThat( importReport.getValidationReport().getWarnings(),
-            hasItem( hasProperty( "warningCode", equalTo( ValidationCode.E1308.name() ) ) ) );
-        assertThat( importReport.getValidationReport().getWarnings(), hasItem( hasProperty( "message", equalTo(
-            "Generated by program rule (`ProgramRulC`) - DataElement `DATAEL00001` is being replaced in event `EVENT123456`" ) ) ) );
-
-        params = fromJson( "tracker/event_update_datavalue.json" );
+
+        params = fromJson( "tracker/programrule/event_update_datavalue_different_value.json" );
         params.setImportStrategy( TrackerImportStrategy.CREATE_AND_UPDATE );
 
         importReport = trackerImportService.importTracker( params );
 
-        assertNoErrors( importReport );
-        warningReports = importReport.getValidationReport().getWarnings();
-        assertEquals( 1, warningReports.size() );
-        assertThat( importReport.getValidationReport().getWarnings(),
-            hasItem( hasProperty( "warningCode", equalTo( ValidationCode.E1308.name() ) ) ) );
-        assertThat( importReport.getValidationReport().getWarnings(), hasItem( hasProperty( "message", equalTo(
-            "Generated by program rule (`ProgramRulC`) - DataElement `DATAEL00001` is being replaced in event `EVENT123456`" ) ) ) );
+        assertHasOnlyErrors( importReport, E1307 );
     }
 
-    private void createProgramRule( char uniqueCharacter, Program program, ProgramRuleActionType actionType )
+    @Test
+    void shouldImportWithWarningWhenDataElementWithDifferentValueIsAssignedByAssignRuleAndOverwriteKeyIsTrue()
+        throws IOException
     {
-        ProgramRule programRuleA = createProgramRule( uniqueCharacter, program );
-        programRuleA.setUid( "ProgramRul" + uniqueCharacter );
-        programRuleService.addProgramRule( programRuleA );
-        ProgramRuleAction programRuleActionShowError = createProgramRuleAction( 'A', programRuleA );
-        programRuleActionShowError.setProgramRuleActionType( actionType );
-        programRuleActionShowError.setContent( "CONTENT" );
-        programRuleActionService.addProgramRuleAction( programRuleActionShowError );
-        programRuleA.getProgramRuleActions().add( programRuleActionShowError );
-        programRuleService.updateProgramRule( programRuleA );
+        assignProgramRule();
+        systemSettingManager.saveSystemSetting( SettingKey.RULE_ENGINE_ASSIGN_OVERWRITE, true );
+        TrackerImportParams params = fromJson( "tracker/programrule/tei_enrollment_completed_event.json" );
+
+        ImportReport importReport = trackerImportService.importTracker( params );
+
+        assertNoErrors( importReport );
```

Comment:
This is something that I am not sure how to tackle.
I need to import something as a setup, and I need to be sure it was correctly imported before I can really test.
Without this assertion, I will probably have the next one failing but it will be quite difficult to understand what went wrong

Reply:
If I'm not mistaken, there are three tests in this class where this strategy is used, and in all of them, the same .json is imported. 
I think we could have a separate test to validate the import is correct, and then in the rest of the tests we assume it works. If not, since then the error message would not be clear, we could have a separate test class and have the json imported during set up.
I mean it, because I think we should aim to assert only one thing, and at the end of the test, asserting in the middle it usually means it's a validation not related to the test. And also in this case, we validate the same in three different tests. Do you agree?

Reply:
Moved to a new test class.
Now it seems much more cleaner
Thank you :)